openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/custom_command_action.hpp

```diff
@@ -32,6 +32,195 @@ struct SpecialAction : public std::integral_constant<int, Value>
+    std::string conditions_name_{""};
+  };
+
+public:
+  SpecialAction() = default;
+  ~SpecialAction() = default;
+
+  // Constructor with no inner object - first object with conditions
+  SpecialAction(
+    const std::string & source_name, const std::string & element_name, int element_index,
+    const std::string & trigger_name_,
+    const std::vector<std::pair<std::string, std::string>> & conditions_list)
+  : source_name_{source_name},
+    inner_{nullptr},
+    core_{std::make_shared<CoreSource>(trigger_name_, conditions_list)},
+    path_{formatPath(element_name, element_index)},
+    log_{core_->comprehensiveLog(path_)}
+  {
+  }
+
+  // Constructor with no inner object - first object without conditions
+  SpecialAction(
+    const std::string & source_name, const std::string & element_name, int element_index,
+    const std::string & trigger_name_)
+  : source_name_{source_name},
+    inner_{nullptr},
+    core_{std::make_shared<CoreSource>(trigger_name_)},
+    path_{formatPath(element_name, element_index)},
+    log_{core_->comprehensiveLog(path_)}
+  {
+  }
+
+  // Constructor with inner object and current source (to be forwarded) - intermediate object
+  SpecialAction(
+    const std::string & source_name, const std::string & element_name, int element_index,
+    SpecialAction const & inner)
+  : source_name_{source_name},
+    inner_{std::make_shared<SpecialAction>(inner)},
+    core_{inner.core_},
+    path_{formatPath(element_name, element_index)},
+    log_{core_->comprehensiveLog(path_)}
+  {
+  }
+
+  // Constructor with inner but without current source -- last object
+  SpecialAction(const std::string & element_name, const SpecialAction & inner)
+  : source_name_{""},
+    inner_{std::make_shared<SpecialAction>(inner)},
+    core_{inner.core_},
+    path_{formatPath(element_name)},
+    log_{core_->comprehensiveLog(path_)}
+  {
+  }
+
+  const char * what() const noexcept { return log_.c_str(); }
+
+private:
+  auto formatPath(const std::string & element_name) -> std::string
+  {
+    if (!inner_) {
+      return element_name;
+    }
+
+    std::stringstream ss("");
+    ss << element_name;
+    if (!inner_->source_name_.empty()) {
+      ss << "." << inner_->source_name_;
+    }
+    ss << inner_->path_;
+    return ss.str();
+  }
+
+  auto formatPath(const std::string & element_name, int element_index) -> std::string
+  {
+    std::stringstream ss("");
+    ss << '.' << element_name << "[";
+    if (inner_ && !inner_->source_name_.empty()) {
+      ss << std::quoted(inner_->source_name_) << "]";
+    } else {
+      ss << element_index << ']';
+    }
+    if (inner_) {
+      ss << inner_->path_;
+    }
+    return ss.str();
+  }
+
+  const std::shared_ptr<const SpecialAction> inner_{nullptr};
+  const std::shared_ptr<const CoreSource> core_{nullptr};
+  const std::string source_name_{""};
+  const std::string path_{""};
+  const std::string log_{"No core source - the path is empty"};
+};
+
```

Comment:
The class works fine, but is overly complex for the simplicity of its role. I was able to refactor the code to make it somewhat better. However, I am still not satisfied.

The problem with this class is that it is hard to imagine from the code the behavior of a gradual expansion of error messages from the depths of the scenario tree to the top level. In particular, each constructor corresponds to one or more specific elements of OpenSCENARIO, but it is hard to tell.

Better code is expected to result from modifying the constructors to accept pointers to element types.
In this case, each element would pass `this` pointer to the `SpecialAction` argument to construct the `SpecialAction` and throw or rethrow it.
In other words, it is necessary to express a clear correspondence between the element type and the respective constructor of SpecialAction.

(Pointers to element types are `Act*`, `ConditionGroup*`, etc. Perhaps a forward declaration is needed.)

However, we do not have enough time to spend on developing this feature, so such modifications will be a future issue.

```suggestion
template <>
class SpecialAction<EXIT_FAILURE> : public std::integral_constant<int, EXIT_FAILURE>,
                                    public std::runtime_error
{
  const std::function<std::string(const std::string &)> make_message;

  const std::string source_name;

  const std::string path;

  template <typename... Ts>
  static auto concatenate(Ts &&... xs)
  {
    std::stringstream ss;
    (ss << ... << xs);
    return ss.str();
  }

  template <typename Container>
  struct comma_separated
  {
    const Container & container;

    const std::function<void(std::ostream &, const typename Container::value_type &)> print;

    template <typename Printer>
    explicit comma_separated(const Container & container, Printer && print)
    : container{container}, print{std::forward<decltype(print)>(print)}
    {
    }

    friend auto operator<<(std::ostream & os, const comma_separated & cs) -> std::ostream &
    {
      auto separtor = "";
      for (const auto & value : cs.container) {
        os << std::exchange(separtor, ", ");
        cs.print(os, value);
      }
      return os;
    }
  };

  template <typename Conditions>
  static auto makeMessage(
    const std::string & trigger_name, const std::string & trigger_path,
    const Conditions & conditions)
  {
    std::stringstream what;

    what << "CustomCommandAction typed " << std::quoted("exitFailure") << " was triggered by the ";

    const auto anonymous = std::any_of(
      conditions.begin(), conditions.end(),
      [](const auto & condition) { return condition.first.empty(); });

    switch (conditions.size()) {
      case 0:
        what << trigger_name << " (" << trigger_path << ")";
        break;

      case 1:
        if (anonymous) {
          what << "anonymous " << trigger_name << " (" << trigger_path << "." << trigger_name
               << "[0]): " << conditions[0].second;
        } else {
          what << trigger_name << " named " << std::quoted(conditions[0].first) << ": "
               << conditions[0].second;
        }
        break;

      default:
        if (anonymous) {
          what << "anonymous " << trigger_name << "s (" << trigger_path << "." << trigger_name
               << "[...]): "
               << comma_separated(conditions, [i = 0](auto & os, const auto & condition) mutable {
                    os << '{' << std::quoted(std::to_string(i++)) << ": " << condition.second
                       << '}';
                  });
        } else {
          what << "named " << trigger_name << "s {"
               << comma_separated(
                    conditions,
                    [](auto & os, const auto & condition) { os << std::quoted(condition.first); })
               << "}: " << comma_separated(conditions, [](auto & os, const auto & condition) {
                    os << '{' << std::quoted(condition.first) << ": " << condition.second << '}';
                  });
        }
        break;
    }

    return what.str();
  }

public:
  SpecialAction()
  : std::runtime_error{
      "CustomCommandAction typed \"exitFailure\" was triggered on an unexpected code path. This is "
      "a simulator bug. Report this to the simulator developer."}
  {
  }

  explicit SpecialAction(
    const std::function<std::string(const std::string &)> & make_message,
    const std::string & source_name, const std::string & path)
  : std::runtime_error{make_message(path)},
    make_message{make_message},
    source_name{source_name},
    path{path}
  {
  }

  // Constructor with no inner object - first object with conditions
  explicit SpecialAction(
    const std::string & source_name, const std::string & element_name, int element_index,
    const std::string & trigger_name,
    const std::vector<std::pair<std::string, std::string>> & conditions = {})
  : SpecialAction{
      [trigger_name, conditions](const auto & trigger_path) {
        return makeMessage(trigger_name, trigger_path, conditions);
      },
      source_name, concatenate('.', element_name, '[', element_index, ']')}
  {
  }

  // Constructor with inner object and current source (to be forwarded) - intermediate object
  explicit SpecialAction(
    const std::string & source_name, const std::string & element_name, int element_index,
    SpecialAction const & inner)
  : SpecialAction{
      inner.make_message, source_name,
      concatenate(
        '.', element_name, '[',
        (inner.source_name.empty() ? std::to_string(element_index) : '"' + inner.source_name + '"'),
        ']', inner.path)}
  {
  }

  // Constructor with inner but without current source -- last object
  explicit SpecialAction(const std::string & element_name, const SpecialAction & inner)
  : SpecialAction{
      inner.make_message, "",
      concatenate(
        element_name, (inner.source_name.empty() ? "" : "." + inner.source_name), inner.path)}
  {
  }
};
```

Reply:
@yamacir-kit 
I've introduced the refactored version of `SpecialAction` you suggested.
I ran local tests - the results of the scenarios are identical.
Perhaps it was indeed redundant to create `CoreSource` structure and its passing, but certainly the `inner_` object was redundant...

I admit that your approach is less complex, I didn't think of this approach, **thank you so much.**


> Better code is expected to result from modifying the constructors to accept pointers to element types.
> In this case, each element would pass this pointer to the SpecialAction argument to construct the SpecialAction and throw or rethrow it.
> In other words, it is necessary to express a clear correspondence between the element type and the respective constructor of SpecialAction.

This approach is also interesting and will indeed introduce more clarity. I will be glad to take care of it in the future.

Changes since the last regression tests are [in this commit](https://github.com/tier4/scenario_simulator_v2/pull/1075/commits/149c6401a9cbfc80afb32fea502cb089bdec88bb). Please let me know if you think re-running the evaluation is necessary. If not, is there anything else I can do for you within this PR?

Reply:
> Please let me know if you think re-running the evaluation is necessary. If not, is there anything else I can do for you within this PR?

I think I need to do it again because the amount of code changes is large... but I ended up merging it due to an operational error.
I think it's probably okay, but I'll test it once just to be sure.